### ISSUE TITLE
Investigate implementation of RoaringArray with Container and Container[]

### ISSUE THREAD

**Issue by richardstartin** [RoaringBitmap/RoaringBitmap#268]
If a bitmap is very high cardinality it makes sense to RLE the keys. If it’s moderate cardinality with more than 4096 keys, a bitmap is a better place to store the keys. This would also make the data structure more tree like allowing an alternative implementation of larger bitmaps.

**Comment by lemire** [comment]
@richardstartin You are correct.

Another option, used by BitMagic, is to have a single array of pointers to containers, and have NULLs where there is no container.

**Comment by richardstartin** [comment]
That's what I imagined when the keys are represented as a bitmap. But when the keys are an array the current configuration is best, and when they would be run compressed I imagine as many arrays as there are runs could be a good option.

**Comment by richardstartin** [comment]
I spent a couple of hours trying to do this and can only conclude that it would be hard to achieve. Partly because it's hard to abstract over the positional bitmap and key based array without cost, so generalising `RoaringArray` requires some thought. Do you expect it to be worth the effort? I can imagine it being worthwhile for aggregations where computing a 16 bit approximation of the aggregation prior to refinement at each level of the tree could allows various efficiencies.

**Comment by lemire** [comment]
I think that the benefits of what you propose can be significant, e.g., for random access.

**Comment by richardstartin** [comment]
Yes, it should behave like a radix tree. I’m not sure run compressed keys would be worth the implementation complexity unless they were part of a deeper tree structure capable of storing things like IP addresses.
